### PR TITLE
Clipping Optimizer

### PR DESCRIPTION
**Pull request by c-h-i-a-m-a-k-a-2**
- Optimizer removes ResolvedClipNodes that are described by a singular ResolvedPathNode and where a clip can be safely applied with out altering the resulting image.
- Documentation: https://docs.google.com/document/d/1OQ3SnGvegSNv2KQVxgHAcTob3p5My1bwWLKGrqXPhFQ/edit?usp=sharing
- Note  MaskingOptimizer (https://github.com/dnfield/vector_graphics/pull/96) needs to get merged first.

### CODE REVIEW

File: packages/vector_graphics_compiler/bin/vector_graphics_compiler.dart [hunk 1]
```diff
@@ -63,7 +68,7 @@ Future<void> main(List<String> args) async {
     }
   }
 
-  if (results['optimize-masks'] == true) {
+  if (results['optimize-masks'] == true || results['optimize-clips'] == true) {
```
Comment (dnfield) [inline packages/vector_graphics_compiler/bin/vector_graphics_compiler.dart:71]: nit: main is getting kind of long, consider moving these to helper methods that are helpfully named e.g. `loadPathOpsIfNeeded(results)` or something like that.

File: packages/vector_graphics_compiler/lib/src/svg/clipping_optimizer.dart [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  List<Node> children = <Node>[];
+  Node parent = Node.empty;
+  bool deleteClipNode = true;
+}
+
+/// Applies and removes trivial cases of clipping.
+class ClippingOptimizer extends Visitor<_Result, Node>
+    with ErrorOnUnResolvedNode<_Result, Node> {
+  ///List of clips to apply.
+  List<Path> clipsToApply = <Path>[];
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/svg/clipping_optimizer.dart:26]:
```suggestion
  final List<Path> clipsToApply = <Path>[];
```

File: packages/vector_graphics_compiler/test/clipping_optimizer_test.dart [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
...0 00-1.558 1.552l-.009 5.845a1.55 1.55 0 001.552 1.55z" fill="#A8DAB5"/><path d="M46.155 53.943l-25.27.04-.04-26.081a2.001 2.001 0 011.993-1.998l21.28-.034a2.001 2.001 0 011.997 1.992l.04 26.08z" fill="#AFCBFA"/><path d="M22.838 25.904a2.006 2.006 0 00-1.91 1.416l-2.065 7.9a3.656 3.656 0 107.314-.01l1.199-9.313-4.538.007z" fill="#669DF7"/><path d="M46.032 27.28a2.008 2.008 0 00-1.918-1.41l-4.538.008 1.229 9.307a3.656 3.656 0 107.314-.011l-2.087-7.894zm-12.554-1.393l-6.102.01-1.199 9.312a3.66 3.66 0 003.664 3.651 3.66 3.66 0 003.65-3.663l-.013-9.31z" fill="#E8F0FD"/><path d="M33.478 25.887l.015 9.31a3.658 3.658 0 007.314-.012l-1.229-9.307-6.1.009z" fill="#669DF7"/><path d="M38.03 53.973l-9.02.013-.015-10.026a1.652 1.652 0 011.65-1.655l5.715-.009a1.654 1.654 0 011.656 1.65l.015 10.027z" fill="#E8F0FE"/><path d="M27.578 11.599h-.006l.052.124.025.061 4.552 10.938a.868.868 0 001.6-.008l4.412-10.873a5.75 5.75 0 00.496-2.47 5.763 5.763 0 00-11.526.143 5.74 5.74 0 00.395 2.085z" fill="#EA4335"/><path d="M32.946 10.998a1.787 1.787 0 100-3.575 1.787 1.787 0 000 3.575z" fill="#A50E0E"/></g><defs><clipPath id="clip0_2630_58398"><path fill="#fff" d="M0 0h64v64H0z"/></clipPath></defs></svg> ''';
```
Comment (dnfield) [inline packages/vector_graphics_compiler/test/clipping_optimizer_test.dart:28]: Consider formatting this in an editor as a multiline indendted string for readability - also consider moving to the end of the file or to the `test_svg_strings.dart` file.

**Review by dnfield — COMMENTED** [review]
Just nits - pinging @jonahwilliams for a review as well

File: packages/vector_graphics_compiler/test/clipping_optimizer_test.dart [hunk 4]
```diff
@@ -0,0 +1,115 @@
+import 'package:test/test.dart';
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/test/clipping_optimizer_test.dart:1]:
```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:test/test.dart';
```

File: packages/vector_graphics_compiler/lib/vector_graphics_compiler.dart [hunk 5]
```diff
@@ -96,10 +98,14 @@ void _encodeShader(
 Future<Uint8List> encodeSvg(
     {required String xml,
     required String debugName,
-    bool? enableMaskingOptimizer}) async {
+    bool? enableMaskingOptimizer,
+    bool? enableClippingOptimizer}) async {
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/lib/vector_graphics_compiler.dart:102]: If we're defaulting to true above, and we're not explicitly passing null, could we remove nullability by making these consistently default to `true`?

File: packages/vector_graphics_compiler/lib/src/svg/clipping_optimizer.dart [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  _Result(this.node);
+
+  final Node node;
+  int childCount = 0;
+  List<Node> children = <Node>[];
+  Node parent = Node.empty;
+  bool deleteClipNode = true;
+}
+
+/// Applies and removes trivial cases of clipping.
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/lib/src/svg/clipping_optimizer.dart:22]: Consider adding an additional paragraph that clarifies what counts as a trivial case

File: packages/vector_graphics_compiler/bin/vector_graphics_compiler.dart (hunk 1 above)
Comment (c-h-i-a-m-a-k-a-2) [reply packages/vector_graphics_compiler/bin/vector_graphics_compiler.dart:71]: fixed.

File: packages/vector_graphics_compiler/test/clipping_optimizer_test.dart (hunk 3 above)
Comment (c-h-i-a-m-a-k-a-2) [reply packages/vector_graphics_compiler/test/clipping_optimizer_test.dart:28]: fixed.

File: packages/vector_graphics_compiler/lib/vector_graphics_compiler.dart (hunk 5 above)
Comment (c-h-i-a-m-a-k-a-2) [reply packages/vector_graphics_compiler/lib/vector_graphics_compiler.dart:102]: fixed.

**Review by jonahwilliams — APPROVED** [review]
LGTM overall, my comments are mostly nits

File: packages/vector_graphics_compiler/lib/src/svg/clipping_optimizer.dart (hunk 6 above)
Comment (c-h-i-a-m-a-k-a-2) [reply packages/vector_graphics_compiler/lib/src/svg/clipping_optimizer.dart:22]: fixed.

**Pull request merged**